### ISSUE TITLE
REST api call - chaining problem

### ISSUE THREAD

**Issue by arvindh5686** [theintern/intern#431]
I am trying to make a REST API call as below and once the call is complete, I want to print "Done". 
But with the below example "Done" is getting printed even before the REST call is complete. 

             return this.remote
                    .then(function() {
                        request('http://www.google.com', function (error, response, body) {
                              if (!error && response.statusCode == 200) {
                                console.log(body) // Show the HTML for the Google homepage.
                              }
                            })
                    })
                    .then(function() {
                        console.log("Done")
                    })
                
Am I missing something here? If this not the right way, could someone please let me know what the right way is.

Thanks.

**Comment by sholladay** [comment]
The first `.then()` call has no way to know when that `request()` thingy is finished. What you need to do is inform the promise chain about what you are doing. There are a few different ways to do that, the cleanest, best way is to `return request(...)`. However, that will only work if `request()` [returns a promise](https://www.npmjs.com/package/request-promise), which will cause the `.then()` promise to follow its lead / depend upon it.

If that's not an option, another would be to create your own promise to wrap the request call.

**Comment by sholladay** [comment]
Here are two great links to help, in case it still isn't clear.

From the beginning:
http://www.html5rocks.com/en/tutorials/es6/promises/#disqus_thread

Pattern variations:
http://pouchdb.com/2015/05/18/we-have-a-problem-with-promises.html

**Comment by arvindh5686** [comment]
Thanks! I was able to solve the problem with deferred promises

**Comment by jason0x43** [comment]
I'm closing this out since it seems to have been resolved.
